backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/ElasticsearchCommonQueryStringPredicate.java

```diff
@@ -0,0 +1,175 @@
+		super( builder );
+		nestedPathHierarchy = builder.firstFieldState.field().nestedPathHierarchy();
+		// Warning: we must use field().absolutePath(), not the keys in the map,
+		// because that key may be a relative path when using SearchPredicateFactory.withRoot(...)
+		fieldPaths = new ArrayList<>( builder.fieldStates.size() );
+		for ( ElasticsearchCommonQueryStringPredicateBuilderFieldState state : builder.fieldStates.values() ) {
+			fieldPaths.add( state.field().absolutePath() );
+		}
+		fieldNameAndBoosts = new ArrayList<>();
+		for ( ElasticsearchCommonQueryStringPredicateBuilderFieldState fieldContext : builder.fieldStates.values() ) {
+			fieldNameAndBoosts.add( fieldContext.build() );
+		}
+		defaultOperator = builder.defaultOperator;
+		queryString = builder.queryString;
+		analyzer = builder.analyzer;
+	}
+
+	@Override
+	protected final JsonObject doToJsonQuery(PredicateRequestContext context, JsonObject outerObject,
+			JsonObject innerObject) {
+		QUERY_ACCESSOR.set( innerObject, queryString );
+		DEFAULT_OPERATOR_ACCESSOR.set( innerObject, defaultOperator );
+
+		JsonArray fieldArray = new JsonArray();
+		for ( JsonPrimitive fieldNameAndBoost : fieldNameAndBoosts ) {
+			fieldArray.add( fieldNameAndBoost );
+		}
+		FIELDS_ACCESSOR.set( innerObject, fieldArray );
+
+		if ( analyzer != null ) {
+			ANALYZER_ACCESSOR.set( innerObject, analyzer );
+		}
+
+		addSpecificProperties( context, outerObject, innerObject );
+
+		queryNameAccessor().set( outerObject, innerObject );
+
+		return outerObject;
+	}
+
+	protected abstract void addSpecificProperties(PredicateRequestContext context, JsonObject outerObject,
+			JsonObject innerObject);
+
+	protected abstract JsonObjectAccessor queryNameAccessor();
+
+	@Override
+	protected List<String> getNestedPathHierarchy() {
+		return nestedPathHierarchy;
+	}
+
+	@Override
+	protected List<String> getFieldPathsForErrorMessage() {
+		return fieldPaths;
+	}
+
+
+	public abstract static class Builder extends AbstractBuilder implements CommonQueryStringPredicateBuilder {
+
+		protected ElasticsearchCommonQueryStringPredicateBuilderFieldState firstFieldState;
+		protected final Map<String, ElasticsearchCommonQueryStringPredicateBuilderFieldState> fieldStates =
+				new LinkedHashMap<>();
+		protected JsonPrimitive defaultOperator = OR_OPERATOR_KEYWORD_JSON;
+		protected String queryString;
+		protected String analyzer;
+
+		Builder(ElasticsearchSearchIndexScope<?> scope) {
+			super( scope );
+		}
+
+		@Override
+		public final void defaultOperator(BooleanOperator operator) {
+			switch ( operator ) {
+				case AND:
+					this.defaultOperator = AND_OPERATOR_KEYWORD_JSON;
+					break;
+				case OR:
+					this.defaultOperator = OR_OPERATOR_KEYWORD_JSON;
+					break;
+			}
+		}
+
+		@Override
+		public final void queryString(String queryString) {
+			this.queryString = queryString;
+		}
+
+		@Override
+		public final FieldState field(String fieldPath) {
```

Comment:
`FieldState` is specific to simple query strings... there's probably something wrong in the type hierarchy?